7zip: translate windows permissions to unix permissions

mostynb opened the pull request:
.7z archives created on windows 7zip can lack unix permission info. In this case, we need to translate the windows permissions into reasonable unix equivalents.

mmatuska commented:
Thank you for the contribution. That should deal with the mode part, but what about fflags? Are FILE_ATTRIBUTE_HIDDEN and FILE_ATTRIBUTE_SYSTEM? Are these stored by 7-zip? We should actually add these as fflags to archive_entry.

See an example in the RAR5 reader:
https://github.com/libarchive/libarchive/blob/5c5a9f2b76ed51f060752b356c9e96ef3aee1baf/libarchive/archive_read_support_format_rar5.c#L1833-L1859

We should probably also fix archive_write_set_format_7zip.c to store the Windows attributes properly - these should be stored in fflags as "readonly", "hidden" and "system":
https://github.com/libarchive/libarchive/blob/5c5a9f2b76ed51f060752b356c9e96ef3aee1baf/libarchive/archive_write_set_format_7zip.c#L1422-L1439

mostynb commented:
Thanks for the pointers. I changed the first commit to use safer permissions, and found that we already had a test case covering this scenario- the test also needed to be updated to match the safer permissions.

I also added the windows file attributes flags as suggested, but using the bitfield function because it seems easier to use and doesn't require a malloc. I haven't looked into the corresponding change for the 7zip writer yet, but I will try to continue later in the week.

mmatuska commented:
Thank you very much for the quick response. We still need a test archive that includes files and directories with FILE_ATTRIBUTE_HIDDEN and FILE_ATTRIBUTE_SYSTEM (we need 4 minimal-sized files and 4 directories inside the archive to test all possible combinations) . The reader reads it and there is already code for checking existing fflags in other tests.

The code in the writer is somewhat ugly:
https://github.com/libarchive/libarchive/blob/5c5a9f2b76ed51f060752b356c9e96ef3aee1baf/libarchive/archive_write_set_format_7zip.c#L1428-L1433

I have now read some specs of the 7z format and we should avoid using magic constants and be more verbose. As what I see we need to change the `struct file {}`, probably by exchanging the `dir` with an `attrs` or passing the hidden and readonly attrs in another way.

mostynb commented:
I copied the RAR5 reader's fflag code, and added a new test which checks {regular,readonly,hidden,system}x{file,directory} cases. I also added some macros to the writier code, to make that a little more readable.
